source/braille.py

```diff
@@ -1688,6 +1689,20 @@ def _doCursorMove(self, region):
 		elif self.buffer is self.messageBuffer and keyboardHandler.keyCounter>self._keyCountForLastMessage:
 			self._dismissMessage()
 
+	def _handleProgressBarUpdate(self, obj):
+		oldTime = getattr(self, "_lastProgressBarUpdateTime", None)
+		newTime = time.time()
+		if oldTime and newTime - oldTime < 1:
+			# Fetching dialog text is expensive, so update at most once a second.
+			return
+		self._lastProgressBarUpdateTime = newTime
+		parent = obj.parent
+		while parent:
+			if parent.role == controlTypes.ROLE_DIALOG:
+				self.handleUpdate(parent)
+				return
+			parent = parent.parent

```

Comment:
Walking parents can be quite expensive. We always walk the parents for the focus and we cache them, but for any other object, this isn't the case.

Have you tested this with progress bars which update quite a lot to see if it affects performance? If it does, consider walking the focus regions, checking for a dialog and just updating the dialog if one is found. While this is wasteful if the progress bar is somewhere else and not inside the dialog, that should be mitigated somewhat by the `isInForeground` check. In contrast, with the current code, progress bars outside of a dialog will pointlessly walk their parents... and I'd say that's far more likely than a progress bar which is in the foreground but not within the focused dialog.

That said, I do agree this code is more readable, so if you aren't seeing a performance hit, I'd be inclined to leave it for now, premature optimisation being the root of all evil and all that. :)